opensearchtransport/opensearchtransport.go

```diff
@@ -258,8 +260,9 @@ func (c *Client) Perform(req *http.Request) (*http.Response, error) {
 				//nolint:errcheck // ignored as this is only for logging
 				buf.ReadFrom(req.Body)
 				req.GetBody = func() (io.ReadCloser, error) {
-					r := buf
-					return io.NopCloser(&r), nil
+					// Return a new reader each time
+					reader := bytes.NewReader(buf.Bytes())
+					return io.NopCloser(reader), nil
```

Comment:
Are we missing a test for this path?

Reply:
There is an existing test for the reader reset behavior on non-compressed request body:
https://github.com/opensearch-project/opensearch-go/blob/41fd58f31b62a0d1edb6f7e506640c79b6a6017f/opensearchtransport/opensearchtransport_internal_test.go#L655-L690

Reply:
But that test is passing without your code change?

Reply:
Notice there are two test cases now, one for retries without compression and one (from this PR) for retries with compression.

The code for uncompressed works correctly and passes the relevant test, both with and without my changes.
The old code does a copy of the unread `buf` into `r` so that `buf` as referenced from the closure remains unread. With my changes, it is a bit more obvious what is being done but it functions the same.
So no, if you revert those two lines, no tests will fail. The bug is only in the code for compressed.

The code for uncompressed was from https://github.com/opensearch-project/opensearch-go/commit/fb7bdd15c1e392734922bece65d3c203a4aefb95 and code for compressed was from https://github.com/opensearch-project/opensearch-go/commit/8cf8bd25e2b7f6fefe3b946f20555272bf8c3c47 but the bug was introduced in https://github.com/opensearch-project/opensearch-go/commit/7ea16f7de5fa5d45a0930619868466f5fc430ee2 with pooled gzip compressors.

Reply:
Understood! Thank you for being patient with me.